### PR TITLE
Improve `Tensor` type hints

### LINKED ISSUE #28457: No type hints for `Tensor.T`
## 🐛 Bug

There are no typehints for `Tensor.T`, so MyPy can't tell it exists.
## To Reproduce

1. Define `test.py` as:
```py
import torch

torch.tensor([[1, 2], [3, 4]]).T
```
1. Run `mypy test.py`:
```
$ mypy test.py
test.py:3: error: "Tensor" has no attribute "T"
Found 1 error in 1 file (checked 1 source file)
```

## Expected behavior
The file should not give a type error.

## Environment

```
$ python collect_env.py 
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-32GB
GPU 1: Tesla V100-PCIE-32GB

Nvidia driver version: 430.26
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip3] numpy==1.17.2
[pip3] torch==1.2.0
[pip3] torchvision==0.4.0
[conda] Could not collect
```
## Additional context

cc @ezyang

### PR DESCRIPTION
**Pull request by henribru**
I've typed some attributes from https://github.com/pytorch/pytorch/blob/ee920b92c4d52c3481ce4fa1594893fde8ccdd93/torch/csrc/autograd/python_variable.cpp#L490 that were not included in the stubs so that MyPy will be aware of them. I made sure to only add those attributes that are mentioned somewhere in the documentation. If there are attributes mentioned in the documentation that are not meant to be part of the public API (or the opposite), please let me know. I've also made sure that attributes that can't be set are typed as read-only properties. If setting `dtype`, `shape`, `device` or `names` directly is not part of the public API, let me know and I'll make them properties as well.

I've also added `__len__`, `__iter__` and `__contains__`, which means MyPy will no longer complain about `len(t)`, `t1 in t2` and `for t1 in t2`.

Shameless plug: I have another typing-related PR here that needs review: https://github.com/pytorch/pytorch/pull/27445

Fixes #28457

### CODE REVIEW

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

File: torch/__init__.pyi.in [hunk 1]
```diff
@@ -83,16 +83,31 @@ class Generator:
...
+    @property
+    def device(self) -> _device: ...
+    @property
+    def T(self) -> Tensor: ...
+    @property
+    def grad_fn(self) -> Optional[Any]: ...
+    @property
+    def ndim(self) -> _int: ...
+    @property
+    def layout(self) -> layout:
```
Comment (fmassa) [inline torch/__init__.pyi.in]: This is an invalid syntax:
```
Oct 24 12:22:38 test_doc_examples (__main__.TestTypeHints) ... torch/__init__.pyi:105: error: invalid syntax
```

**Review by fmassa — CHANGES_REQUESTED** [review]
Test failures seems related https://circleci.com/gh/pytorch/pytorch/3336395?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: torch/__init__.pyi.in (hunk 1 above)
Comment (henribru) [reply torch/__init__.pyi.in]: Sorry about that, should be fixed now.

**Comment by henribru** [comment]
https://github.com/python/mypy/issues/4146 is causing problems it seems. I'll have a look.

**Review by fmassa — APPROVED** [review]

**Pull request closed without merging**